File: exercise/ts/week02-object-calisthenics/exercise2/src/BookStore.ts
import { Book } from './Book';

export class BookStore {
    private inv: Book[] = [];

    addBook(title: string | null, author: string | null, copies: number): void {
        if (title !== null && author !== null && copies > 0) {
            let foundBook: Book | null = null;
            for (const book of this.inv) {
                if (book.getTitle() === title && book.getAuthor() === author) {
                    foundBook = book;
                    break;
                }
            }
            if (foundBook !== null) {
                foundBook.addCopies(copies);
            } else {
                this.inv.push(new Book(title, author, copies));
            }
        }
    }

    sellBook(title: string, author: string, copies: number): void {
        for (const book of this.inv) {
            if (book.getTitle() === title && book.getAuthor() === author) {
                book.removeCopies(copies);
                if (book.getCopies() <= 0) {
                    this.inv = this.inv.filter(b => b !== book);
                }
                break;
            }
        }
    }
}

File: exercise/java/week02-object-calisthenics/exercise2/src/main/java/BookStore.java
import java.util.ArrayList;

public class BookStore {
    private ArrayList<Book> inv = new ArrayList<>();

    public void addBook(String title, String author, int copies) {
        if (title != null && author != null && copies > 0) {
            Book foundBook = null;
            for (Book book : inv) {
                if (book.getTitle().equals(title)
                        && book.getAuthor().equals(author)) {
                    foundBook = book;
                    break;
                }
            }
            if (foundBook != null) {
                foundBook.addCopies(copies);
            } else {
                inv.add(new Book(title, author, copies));
            }
        }
    }

    public void sellBook(String title, String author, int copies) {
        for (Book book : inv) {
            if (book.getTitle().equals(title)
                    && book.getAuthor().equals(author)) {
                book.removeCopies(copies);
                if (book.getCopies() <= 0) {
                    inv.remove(book);
                }
                break;
            }
        }
    }
}
